src/shared/sam/debugger/awsSamDebugger.ts

```diff
@@ -181,6 +182,24 @@ export class SamDebugConfigProvider implements vscode.DebugConfigurationProvider
             templateResource?.Properties?.Runtime ??
             getDefaultRuntime(getRuntimeFamily(editor?.document?.languageId ?? 'unknown'))
 
+        // Merge env vars from launch.json + template.yaml.
+        // Note: launch.json takes precedence.
```

Comment:
Are you building the temporary template off of this combined env var object? If so, there's no reason to build a separate env var file: https://github.com/aws/aws-toolkit-vscode/pull/1100/files#diff-479be7d749f50f1b4db77dd5171fcb36R205 and you don't need the logic below (since you'll already have combined them).

That being said, that goes against how SAM works (doesn't break but will only overwrite env vars if they're present in the template), so the question is how badly do we want to abstract away SAM?

Reply:
> but will only overwrite env vars if they're present in the template

Not sure I follow: I thought we are required to ignore env vars not present in template.yaml. Or what do you mean?

Reply:
If I'm reading this correctly, you're:
* creating a new template construct, containing both env vars from the template and the launch config
* removing the ones that weren't in the original template
* creating a template with these values
* copying the launch config's env vars to a separate .json file and loading that into your SAM commands

There's at least one redundancy there. If you had zero processing on this aside from taking the env vars in the launch config and saving them to a file for SAM to use (which is what we did before), it would override the env vars that matched the template and throw out the rest. By building your struct, you're doing all of the hard work that SAM's doing (left-join with configs taking precedent) and yet you still have the temporary file.

At the very least, I'd eliminate any manual joins that you're doing....unless:

The other thing I'm bringing up is that SAM's behavior is to throw out the unused overrides silently. I'm fairly sure that our current behavior is to stop execution and alert the user? Which is a bit extreme but prevents them from falling into an unexpected state (env vars that they manually entered into the config aren't there). While you're here, we have a few options:

1. Pass the env var JSON through with no processing and let SAM do its thing (which could result in user confusion but lets SAM be SAM)
2. Block execution like we're doing and potentially frustrate the user but ensure they know what's happening
3. Do what you're currently doing with the manual joins and merge everything in the template file we create (and not provide the separate env vars JSON), which means that the user always gets the env vars they entered, but it's not exactly the native SAM experience.

Reply:
> SAM's behavior is to throw out the non-overriding overrides from the env var JSON silently.

That wasn't happening , so I did the left-join. But now I understand why: I was copying  `lambda..evironment` to the `input-template.yaml` and shouldn't do that.

> our current behavior is to stop execution and alert the user?

It's just a warning/informational, doesn't stop anything.

> 3\. Do what you're currently doing with the manual joins and merge everything in the template file we create (and not provide the separate env vars JSON), which means that the user always gets the env vars they entered, but

current approach:

- advantage 1: it's testable in the CONFIG phase, thus we get much more coverage
- advantage 2: env var logic is in one place instead of spread across `env-vars.json` + template.
- disadvantage (as you mentioned): if/when we eliminate the temporary `input-template.yaml`, the logic needs to be reworked again.

Disadvantage is strong so I will rework this now.